From 87f44d1ae6916048c3dc3b16de66005c10bffc1d Mon Sep 17 00:00:00 2001
From: Daniel Leung <daniel.leung@intel.com>
Date: Tue, 2 Oct 2012 16:17:24 -0700
Subject: Skip non-writable rfkill nodes

Sometimes the platform has a lower numbered rfkill node that is related
to bluetooth, but does not directly control the rfkill in the bluetooth
driver. For example, Lenovo ThinkPad has a bluetooth radio control in
ACPI. However, the USB bluetooth module has actual control for the
software radio kill. This skips the first rfkill node, so that bluedroid
can talk to the correct node.

Change-Id: Ic9c6fef08d2383ce20a10d72833e54b73746c850
Signed-off-by: Daniel Leung <daniel.leung@intel.com>
---
 bluedroid/bluetooth.c |    7 +++++++
 1 files changed, 7 insertions(+), 0 deletions(-)

diff --git a/bluedroid/bluetooth.c b/bluedroid/bluetooth.c
index ab9daf5..dbfddd3 100644
--- a/bluedroid/bluetooth.c
+++ b/bluedroid/bluetooth.c
@@ -75,6 +75,13 @@ static int init_rfkill() {
             close(fd);
             if (sz >= 9 && memcmp(buf, "bluetooth", 9) == 0) {
                 asprintf(&rfkill_state_path, "/sys/class/rfkill/%s/state", de->d_name);
+                if (access(rfkill_state_path, W_OK)) {
+                    // 'state' has to be writable. skip the ones that are not.
+                    ALOGW("open(%s) failed: %s (%d)\n", rfkill_state_path, strerror(errno), errno);
+                    free(rfkill_state_path);
+                    rfkill_state_path = NULL;
+                    continue;
+                }
                 break;
             }
         }
-- 
1.7.4.1

